shell/server/core.js

```
@@ -0,0 +1,220 @@
+    }
+  },
+  readAllNotifications: function () {
+    if (!Meteor.userId()) {
+      throw new Meteor.Error(403, "User not logged in.");
+    }
+    Notifications.update({userId: Meteor.userId()}, {$set: {isUnread: false}}, {multi: true});
+  }
+});
+
+NotificationHandle.prototype.close = function () {
+  var self = this;
+  return inMeteor(function () {
+    tryDropNotification(self.notificationId);
+  });
+};
+
+NotificationHandle.prototype.save = function () {
+  var self = this;
+  return inMeteor(function () {
+    var sturdyRef = new Buffer(Random.id(20));
+    var hashedSturdyRef = Crypto.createHash("sha256").update(sturdyRef).digest("base64");
+    ApiTokens.insert({
+      _id: hashedSturdyRef,
+      frontendRef: {
+        notificationHandle: self.notificationId
+      }
+    });
+    return {sturdyRef: sturdyRef};
+  });
+};
+
+SandstormCoreImpl.prototype.restore = function (sturdyRef) {
+  var self = this;
+  return inMeteor(function () {
+    var hashedSturdyRef = Crypto.createHash("sha256").update(sturdyRef).digest("base64");
+    var token = ApiTokens.findOne(hashedSturdyRef);
+    if (!token) {
+      throw new Error("No token found to restore");
+    }
+    if (token.frontendRef) {
+      var notificationId = token.frontendRef.notificationHandle;
+      return {cap: makeNotificationHandle(notificationId)};
+    } else if (token.objectId) {
+      return waitPromise(openGrain(self.grainId).supervisor.restore(token.objectId).catch(function (err) {
+        if (shouldRestartGrain(err, 0)) {
+          return openGrain(grainId, true).supervisor.restore(token.objectId);
+        }
+      }));
+    } else {
+      throw new Error("Unknown token type.");
+    }
+  });
+};
+
+var dropToken = function (grainId, sturdyRef) {
+  var hashedSturdyRef = Crypto.createHash("sha256").update(sturdyRef).digest("base64");
+  var token = ApiTokens.findOne({_id: hashedSturdyRef});
+  if (!token) {
+    return;
+  }
+  if (token.frontendRef) {
```

In this if/else, you're assuming that all frontendRefs are notificationHandles and that all objectIds are wakeLockNotifications, but these are both unions that could contain other things, so please check and throw an error on unsupported types.